arch/x86/src/boundary/boundary.rs

```diff
@@ -0,0 +1,215 @@
+            state.write_stack(1, ret1, accessible_memory_start, app_brk)?;
+            state.write_stack(2, ret2, accessible_memory_start, app_brk)?;
+            state.write_stack(3, ret3, accessible_memory_start, app_brk)?;
+        }
+
+        Ok(())
+    }
+
+    unsafe fn set_process_function(
+        &self,
+        accessible_memory_start: *const u8,
+        app_brk: *const u8,
+        state: &mut Self::StoredState,
+        upcall: FunctionCall,
+    ) -> Result<(), ()> {
+        // Our x86 port expects upcalls to be standard cdecl routines. We push args and return
+        // address onto the stack accordingly.
+        //
+        // Upcall arguments are written directly into the existing stack space (rather than
+        // being pushed on top). This is safe to do because:
+        //
+        // * When the process first starts ESP is initialized to `app_brk - 16`, giving us exactly
+        //   enough space for these arguments.
+        // * Otherwise, we assume the app is currently issuing a `yield` syscall. We re-use the
+        //   stack space from that syscall. This is okay because `yield` doesn't return anything.
+        //
+        // Safety: Caller of this function has guaranteed that the memory region is valid.
+        unsafe {
+            state.write_stack(0, upcall.argument0 as u32, accessible_memory_start, app_brk)?;
+            state.write_stack(1, upcall.argument1 as u32, accessible_memory_start, app_brk)?;
+            state.write_stack(2, upcall.argument2 as u32, accessible_memory_start, app_brk)?;
+            state.write_stack(3, upcall.argument3 as u32, accessible_memory_start, app_brk)?;
+
+            state.push_stack(state.eip as u32, accessible_memory_start, app_brk)?;
+        }
+
+        // The next time we switch to this process, we will directly jump to the upcall. When the
+        // upcall issues `ret`, it will return to wherever the yield syscall was invoked.
+        state.eip = upcall.pc as u32;
+
+        Ok(())
+    }
+
+    unsafe fn switch_to_process(
+        &self,
+        accessible_memory_start: *const u8,
+        app_brk: *const u8,
+        state: &mut Self::StoredState,
+    ) -> (ContextSwitchReason, Option<*const u8>) {
+        // Sanity check: don't try to run a faulted app
+        if state.exception != 0 || state.err_code != 0 {
+            let stack_ptr = state.esp as *mut u8;
+            return (ContextSwitchReason::Fault, Some(stack_ptr));
+        }
+
+        let mut err_code = 0;
+        let int_num = unsafe { super::switch_to_user(state, &mut err_code) };
+
+        let reason = match int_num as u8 {
+            0..=0x1f => {
+                state.exception = int_num as u8;
+                state.err_code = err_code;
+                ContextSwitchReason::Fault
+            }
+
+            SYSCALL_VECTOR => {
+                let num = state.eax as u8;
+
+                // Syscall arguments are passed on the stack using cdecl convention.
```

Comment:
Two reasons. The most important one was to make syscalls "look like" cdecl, which in turn allows upcall handlers to actually _be_ regular cdecl functions.

Second minor reason is because x86 simply has fewer general registers.

Main documentation for the user/kernel boundary is currently in _boundary/mod.rs_, but we're totally open to suggestions for a more central way to document this area.

Reply:
I feel that the question is sufficiently answered by the doc comments in _boundary/mod.rs_, so I'm marking this as resolved. Please feel free to reopen if you feel the comments are insufficient or if you think there's a better place to store this documentation.